src/dbnode/persist/fs/msgpack/decoder.go

```diff
@@ -328,7 +349,11 @@ func (dec *Decoder) decodeIndexEntry() schema.IndexEntry {
 		return indexEntry
 	}
 
-	indexEntry.EncodedTags, _, _ = dec.decodeBytes()
+	if dec.byteReader != nil {
```

Comment:
Hm why does line 337 use `bytesPool == nil` and this if/else use `dec.byteReader != nil`?  Shouldn't it be the same condition?

Reply:
Maybe make this a helper function? Like `func (dec *Decoder) decodeBytesDirect() bool`?  Then you can do:
```go
if dec.decodeBytesDirect() {
  //...
} else {
  //...
}
```

Reply:
its a remnant of a previous implementation I had where the decoder stored a checked bytes pool and you had to make decisions based on whether or not various interfaces were implemented...

I think for now the right thing to do is to check if the bytes pool is passed, it feels more explicit. If you try and use the decoder the wrong way (pass nil for bytes pool, but also try and use a file instead of a stream backed by a byte slice) the decoder will give you an appropriate error `errorCalledDecodeBytesWithoutByteStreamDecoder`

Reply:
Basically if the bytes pool is present the caller is signaling that they want to use the implementation that doesn't involve referencing the backing byte slice implementation.

I tried 3-4 different approaches and this is the one that came out the cleanest, but I'm open to suggestions